### ISSUE TITLE
[Bug] Reverse input channels on BHWC input

### ISSUE THREAD

**Issue by tersekmatija** [openvinotoolkit/openvino#7672]
##### System information (version)
- OpenVINO=> 2021.4
- Operating System / Platform => Linux / Google Colab
- Problem classification => Model Optimizer (mo.py)

##### Detailed description
Hey, I've used tflite2tensorflow to convert the tflite Selfie Segmentation Landscape model from MediaPipe (https://github.com/google/mediapipe/tree/master/mediapipe/modules/selfie_segmentation) to ONNX. I compared the inference results from the MediaPipe and ONNX and they seem OK. However, I want to now use mo.py to generate OpenVINO bin and xml. I use the following command:

`
python /opt/intel/openvino_2021/deployment_tools/model_optimizer/mo_onnx.py \
--input_model  /content/model_float32.onnx \
--model_name selfie_segmentation_landscape_ric \
--data_type FP16 \
--scale 255 \
--input_shape [1,144,256,3] \
--reverse_input_channels \
--output_dir /content/output/
`

which gives the following warning:
`
[ WARNING ]  Network has 1 inputs overall, but only 0 of them are suitable for input channels reversing.
Suitable for input channel reversing inputs are 4-dimensional with 3 channels
All inputs: {'input_1': [1, 144, 256, 3]}
Suitable inputs {}
`

As I understand from the documentation, the command should work and not produce the warning, if I specify the input_shape which I do. Any idea what might go wrong?

In the attachment you can find the ONNX file.

Thank you in advance!

[model_float32.zip](https://github.com/openvinotoolkit/openvino/files/7234631/model_float32.zip)

**Comment by Iffa-Meah** [comment]
Hi,

I had tested your model and it seems that your model has issues even before the conversion.
I tried to convert your model without the reverse parameter but it doesn't make any differences. Not sure if you had noticed, your model actually returns the shape values of float64 instead of float32.

**Please note that your model is not listed in [the supported topology list](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow.html). Any model and topology that is not listed is considered unsupported by OpenVINO and issues are expected.**




![onnx_benchmark](https://user-images.githubusercontent.com/79643476/135194433-e0ddd68c-600d-4cf2-b4b9-3a2a3ea1ed70.png)
![convert_noreverse](https://user-images.githubusercontent.com/79643476/135194503-598d6d0c-858e-4e9f-8c6f-6409e5edfd6e.png)
3
![IR_benchmark](https://user-images.githubusercontent.com/79643476/135194523-09abd27a-bb3a-43a2-8a1f-baa21543455d.png)
2

**Comment by Iffa-Meah** [comment]
I believe that warning came up because you used the --reverse_input_channels. The warning mentioned about, the criteria of input channel that is suitable to be used with the reverse parameter, in which not really suitable for your model.

I tried converting your model without that parameter and I didn't get that warning. Instead, it prompting about your cons node returns shape value of float 64 and somewhat forced to be casted into float32. 

Usually warnings prompted good to be kept in mind but could be ignored as long as it does not produce any error.

![convert_noreverse](https://user-images.githubusercontent.com/79643476/135206306-e5e9eb50-64d0-4c3d-bcae-5f377c16fa0d.png)

**Comment by tersekmatija** [comment]
Yeah @Iffa-Meah , it happens because I use `--reverse_input_channels` on a HWC input and gives the error that no suitable inputs for channel reversing have been found.  And without it does not produce the reverse input channels warning.

Is reverse_input_channels only supported on CHW inputs, or why my input is not supported? If so, this info could be added to the documentation here (https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_Converting_Model_General.htm)

**Comment by Iffa-Meah** [comment]
The --reverse_input_channels should be used depending on the input data for your application, which can be of RGB or BRG color input order. 

For example, Inference Engine samples load input images in the BGR channels order. However, the model may be trained on images loaded with the opposite order (for example, most TensorFlow* models are trained with images in RGB order). 

In this case, inference results using the Inference Engine samples may be incorrect.
The solution is to provide --reverse_input_channels command line parameter. Taking this parameter, the Model Optimizer performs first convolution or other channel-dependent operation weights modification so these operations output will be like the image is passed with RGB channels order.

In short, it switches the input channels order from RGB to BGR (or vice versa) and should be applied to original inputs of the model **if and only if a number of channels equals 3**.

You may refer [here](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_Converting_Model_General.html#when_to_reverse_input_channels).

**Comment by tersekmatija** [comment]
Hey @Iffa-Meah , I understand all that, but this is not what I am asking at all.

So, my application is reading images in BGR order, and my model takes the input as RGB. Due to the nature of my application, it is not recommended to do any preprocessing (like BGR to RGB switch) in the code, that's why I want to use `--reverse_input_channels`.

But the issue is I can't, because I get the no suitable channels warning, and consequently the order is not reversed. 

I think the issue is that the input of my model is BHW**C**, so [1, 144, 256, **3**], so the **number of color channels equals 3**, but they are located in the last dimension. However, this does not happen to me when dealing with models that take B**C**HW as input (for example [1, 3, 144, 256]). So my question specifically is, does the input shape need to be B**C**HW in order to use the `--reverse_input_channels`? If not and BHWC is also accepted, then I believe mo.py shouldn't give a warning for my input of shape [1, 144, 256, **3**].

**Comment by jgespino** [comment]
Hi @tersekmatija 

I don't believe the warning is due to the model's input being BHWC, I was able to convert and reverse the input channels on a model from the [Open Model Zoo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/public/fcrn-dp-nyu-depth-v2-tf) that was also BHWC format. Let me check with the development team and see if there is anything specific in your model that is causing the warning.

Regards,
Jesus

Ref. 67222

**Comment by bilalsoomro** [comment]
Hi. I am facing the same issue.

My model has the BHWC input shape of { 1, 224, 224, 3 } and it is trained with the RGB channel order. When I tried to use the `--reverse_input_channels` option during conversion, it gives me the same error.

> [ WARNING ]  Network has 1 inputs overall, but only 0 of them are suitable for input channels reversing.
Suitable for input channel reversing inputs are 4-dimensional with 3 channels
All inputs: {'input:0': [1, 224, 224, 3]}
Suitable inputs {}

**Comment by jgespino** [comment]
Hi @bilalsoomro 

Would it be possible for you to share your model for us to take a look?

Regards,
Jesus

**Comment by bilalsoomro** [comment]
Hi @jgespino, here's the [link ](https://drive.google.com/file/d/1dGxzY9nldpXIdvrH-e8Of3tMzXllVnDb/view?usp=sharing)to the model.

**Comment by jgespino** [comment]
Hi @bilalsoomro

I see the warning with your model as well, I'll need to check with the development team for additional details. In the meantime, I am able to run inference with your model in onnx and IR format using the benchmark_app. Have you tried running inference on your model and do you obtain the expected results? 

Regards,
Jesus

**Comment by jgespino** [comment]
Hi @tersekmatija @bilalsoomro 

The warning is shown as the model optimizer determines the C channel based on the default layout (NCHW for ONNX models). The development team is working on implementing new logic, once I have the PRs I will share with you to test. 

Regards,
Jesus

**Comment by bilalsoomro** [comment]
Hi @jgespino , the model does work for me once I convert it and use the shape that my model expects. Just wanted to share the error as by default OpenVino uses BCHW and it wasn't able to reverse the channels with that flag.

**Comment by jgespino** [comment]
@bilalsoomro Great, thanks for confirming. 

@tersekmatija Does your model work as expected too? There is a PR to address the issue but has not been merged/reviewed yet.
https://github.com/openvinotoolkit/openvino/pull/9046

**Comment by jgespino** [comment]
Hi @bilalsoomro  @tersekmatija 

PR https://github.com/openvinotoolkit/openvino/pull/9355 was merged and I was able to convert your models using the following command on the master branch. Note, the layout needs to be specified `--layout nhwc` . Please give it a try and let me know if you have any questions.

```
mo --input_model model_float32.onnx --data_type FP16 --scale 255 --input_shape [1,144,256,3] --reverse_input_channels --layout nhwc
Model Optimizer arguments:
Common parameters:
	- Path to the Input Model: 	/home/pse/Downloads/model_float32.onnx
	- Path for generated IR: 	/home/pse/Downloads/.
	- IR output name: 	model_float32
	- Log level: 	ERROR
	- Batch: 	Not specified, inherited from the model
	- Input layers: 	Not specified, inherited from the model
	- Output layers: 	Not specified, inherited from the model
	- Input shapes: 	[1,144,256,3]
	- Source layout: 	Not specified
	- Target layout: 	Not specified
	- Layout: 	nhwc
	- Mean values: 	Not specified
	- Scale values: 	Not specified
	- Scale factor: 	255.0
	- Precision of IR: 	FP16
	- Enable fusing: 	True
	- Enable grouped convolutions fusing: 	True
	- Move mean values to preprocess section: 	None
	- Reverse input channels: 	True
	- Use legacy API for model processing: 	False
	- Use the transformations config file: 	None
ONNX specific parameters:
	- OpenVINO runtime found in: 	/opt/intel/openvino_2022/python/python3.6/openvino
OpenVINO runtime version: 	2022.1.custom_master_f255c195c5fd9354dfd86379667f99f2f63776cb
Model Optimizer version: 	2022.1.custom_master_f255c195c5fd9354dfd86379667f99f2f63776cb
[ WARNING ]  
Detected not satisfied dependencies:
	networkx: installed: 2.5.1, required: ~= 2.6

Please install required versions of components or use install_prerequisites script
/home/pse/new_build/openvino/tools/mo/install_prerequisites/install_prerequisites_onnx.sh
Note that install_prerequisites scripts may install additional components.
[ SUCCESS ] Generated IR version 11 model.
[ SUCCESS ] XML file: /home/pse/Downloads/model_float32.xml
[ SUCCESS ] BIN file: /home/pse/Downloads/model_float32.bin
[ SUCCESS ] Total execution time: 10.26 seconds. 
[ SUCCESS ] Memory consumed: 118 MB. 
```

```
mo --input_model mobilenetv3_fp32.onnx --batch 1 --data_type FP16 --reverse_input_channels --layout nhwc
Model Optimizer arguments:
Common parameters:
	- Path to the Input Model: 	/home/pse/Downloads/mobilenetv3_fp32.onnx
	- Path for generated IR: 	/home/pse/Downloads/.
	- IR output name: 	mobilenetv3_fp32
	- Log level: 	ERROR
	- Batch: 	1
	- Input layers: 	Not specified, inherited from the model
	- Output layers: 	Not specified, inherited from the model
	- Input shapes: 	Not specified, inherited from the model
	- Source layout: 	Not specified
	- Target layout: 	Not specified
	- Layout: 	nhwc
	- Mean values: 	Not specified
	- Scale values: 	Not specified
	- Scale factor: 	Not specified
	- Precision of IR: 	FP16
	- Enable fusing: 	True
	- Enable grouped convolutions fusing: 	True
	- Move mean values to preprocess section: 	None
	- Reverse input channels: 	True
	- Use legacy API for model processing: 	False
	- Use the transformations config file: 	None
ONNX specific parameters:
	- OpenVINO runtime found in: 	/opt/intel/openvino_2022/python/python3.6/openvino
OpenVINO runtime version: 	2022.1.custom_master_f255c195c5fd9354dfd86379667f99f2f63776cb
Model Optimizer version: 	2022.1.custom_master_f255c195c5fd9354dfd86379667f99f2f63776cb
[ WARNING ]  
Detected not satisfied dependencies:
	networkx: installed: 2.5.1, required: ~= 2.6

Please install required versions of components or use install_prerequisites script
/home/pse/new_build/openvino/tools/mo/install_prerequisites/install_prerequisites_onnx.sh
Note that install_prerequisites scripts may install additional components.
[ SUCCESS ] Generated IR version 11 model.
[ SUCCESS ] XML file: /home/pse/Downloads/mobilenetv3_fp32.xml
[ SUCCESS ] BIN file: /home/pse/Downloads/mobilenetv3_fp32.bin
[ SUCCESS ] Total execution time: 6.38 seconds. 
[ SUCCESS ] Memory consumed: 156 MB.

```
